### ISSUE TITLE
start play over finished listener leads to problems

### ISSUE THREAD

**Issue by der-lange-christian** [caprica/vlcj#386]
I try to play one song after one song finished. I tried it with vlcj 3.1.0 and changed to vlcj 3.8.0 but it become more problematic than before. It is just stops working under vlcj3.1.0 after a few starts but come then to a problem in java (when you interested I can send a stacktrace). With 3.8.0 I get an error earlier (normally less than 20 starts) It just stuck without an error (sometimes with error). The program is printed below. The track I use is like the name says only 2 seconds long so I get fast many starts.

Do you need more information? You can also increase the number of "Testers" than the error comes faster. Do I something fundamental wrong? I start the player in a new Thread because I've read somewhere starting in the listener is not a good idea.

- vlc --version :
   + VLC media player 2.2.0-pre2
- java --version: 
   + java version "1.8.0_45"
   + Java(TM) SE Runtime Environment (build 1.8.0_45-b14)
   + Java HotSpot(TM) 64-Bit Server VM (build 25.45-b02, mixed mode)
- uname -a:
   + Linux ubuntu 3.16.0-28-generic #38-Ubuntu SMP Fri Dec 12 17:37:40 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

```java
import java.io.File;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import uk.co.caprica.vlcj.player.MediaPlayer;
import uk.co.caprica.vlcj.player.MediaPlayerEventAdapter;
import uk.co.caprica.vlcj.player.MediaPlayerFactory;


public class RepeatingPlayingMinimal {

    private static volatile AtomicInteger counter = new AtomicInteger(0);
    
    private static final MediaPlayerFactory factory = new MediaPlayerFactory("--no-video-title-show");
    private static final File song = new File("test-data/shortTrack2sec.mp3");

    public static void main(String[] args) {
        // MediaPlayers will not be garbish collected
        final List<MediaPlayer> list = new ArrayList<>();
        
        System.out.println("start playing\n\n");
        for (int i = 0; i < 1; ++i) {
            
            final String name = "Tester: " + i;
            final int number = i;
        
            list.add(i, factory.newEmbeddedMediaPlayer());
            list.get(i).addMediaPlayerEventListener(new MediaPlayerEventAdapter() {
                
                    @Override
                    public void finished(MediaPlayer mp) {
                        System.out.println(name + " - finished");
                        Tester tester = new RepeatingPlayingMinimal.Tester(name, list.get(number));
                        Thread t = new Thread(tester);
                        t.setName(tester.name);
                        t.start();
                    }
                    
                    @Override
                    public void timeChanged(MediaPlayer mediaPlayer, long time) {
                        //System.out.println(name + " - timeChanged: " + time);
                    }
            });
            
            Tester tester = new RepeatingPlayingMinimal.Tester(name, list.get(i));
            Thread t = new Thread(tester);
            t.setName(name);
            t.start();
        }

        try {
            Thread.sleep(Integer.MAX_VALUE);
        } catch (InterruptedException ex) {
            ex.printStackTrace();
        }
    }

    private static class Tester implements Runnable {
        
        final MediaPlayer mediaPlayer;
        final String name;
        
        Tester(String name, MediaPlayer mediaPlayer) {
            this.mediaPlayer = mediaPlayer;
            this.name = name;
        }

        @Override
        public void run() {
            try {
                System.out.println(name + " - start: " + counter.incrementAndGet());
                mediaPlayer.playMedia(new String(song.getAbsolutePath().getBytes("UTF-8")));
                
            } catch (UnsupportedEncodingException ex) {
                ex.printStackTrace();
            }
        }

    }
}
```

**Comment by caprica** [comment]
So here what is happening is that when you receive the finished event for the first media and the vlcj listener fires, you are very quickly calling back into VLC to play your next media. The problem is that (sometimes, it's a race condition) you are calling back into VLC while VLC is still processing the event callback for that first finished event. i.e. the finished event processing has not finished, and you are calling back into VLC.

vlcj tries to be clever (too clever, probbably) by offloading the event processing to a separate thread - what this does is in theory is give the VLC event callback a bit more time to finish. In practice this is unreliable.

You could say vlcj should not offload event processing a thread, so the native event processing is therefore synchronous. This is a fair comment to be honest. However, if vlcj were changed to synchronously process events in this way then you are almost certainly guaranteed a failure for your specific problem because it is simply not allowed to call back into VLC during a VLC event callback. This is basically what you have seen with your problem description anyway.

So there's no nice solution for this unfortunately, I wish there were.

To do what you want, you are really supposed to use the MediaList functionality instead of using these events. These events are more suited for where someone is pressing a button in a UI to start media.

**Comment by caprica** [comment]
By the way, are you sure in your code you are pinning object references correctly? You must keep hard references to all vlcj objects to prevent them being garbage collected, otherwise strange behaviour (crashes) can occur.

You should enable the vlcj logging and make sure no finalize() methods are being called.

**Comment by der-lange-christian** [comment]
Thank you very much for the explanation, nice to know what the problem is or at least to have an idea.

I enabled the logging and saw that no finalize() was called. The play use-case was only an example I had also problems when I tried to simulate a fade in and fade out. Is there any possibility to find out if vlc is still working with other events? (it's a mess with race conditions - I thought it would be good to have a method I could call, but when I called another event handler could started working - why such a condition is so seldom when it could happen in every setVolume()?)

Can the same happen with getVolume() and setVolume()? Is it dangerous to call such methods to often? Or to call them while a event handler is working? Is there any native solution for fading in vlc? For me it is all the same issue so I hope the questions are not to off topic.

**Comment by caprica** [comment]
You can not call back into LibVLC while it is processing an event, it's as simple as that.

You can not call back into LibVLC if it is already processing a prior request.

This latter point is not always obvious. Whilst your native API call is synchronous, lots of asynchronous things can go on in the background as a result of those native API calls. This can cause you problems, so yes like too many set/get volume *potentially* can cause a problem (but in this particular case I doubt it).

Just try calling playMedia() 1000's of times in a tight loop. It will probably crash.

**Comment by caprica** [comment]
There is nothing in the LibVLC API to help you fade anything, no.

**Comment by caprica** [comment]
As for why it happens sometimes but not others, well that is inherent in a race isn't it. But more particularly, get/set volume doesn't really do much. Whereas playing media kicks off all sorts of decoder and effect plugins, so much more scope for things to go wrong or slowly.

**Comment by der-lange-christian** [comment]
And set/get Volume while starting a media is also problem? When I use the MediaList is set/get Volume still a problem (e.g. when a new mp3 file is starting)?

**Comment by caprica** [comment]
I don't know what you mean. Get/set volume before starting media is unreliable. Setting volume can be asynchronous, so getting volume immediately after setting it may lead to wrong results.

You should check latest vlc-3.0.0-git and latest vlcj-3.9.0-snapshot for improved volume handling (there are now native events for volume changes and mute changes).

**Comment by der-lange-christian** [comment]
When I use more than one MediaPlayer are the problems restricted per instance or is it a problem when I start a mp3 file in mediaPlayerOne and do the same in mediaPlayerTwo that this both interaction lead to race condition?

**Comment by der-lange-christian** [comment]
I think I do not understand up to now which are dangerous methods and which are not. When I use addMediaPlayerEventListener and need very long in this could it be a problem when I call from another thread in the same time a set/getVolume?

**Comment by caprica** [comment]
I don't know what else I can say - the only "rule" is that you can not call back into LibVLC while it is processing a previous call, or while you are inside an event callback. So if you want to be 100% reliable, don't do that.

**Comment by der-lange-christian** [comment]
To take care that I'm not starting within an event is still working this code could be a "work around"? Just add 500 ms to let vlc handle his internals? Could you give an advice for the length of the time?
```java
                Thread.sleep(500);
                System.out.println(name + " - start: " + counter.incrementAndGet());
                mediaPlayer.playMedia(new String(song.getAbsolutePath().getBytes("UTF-8")));
```

**Comment by caprica** [comment]
Well that will probably work 99% of the time. The problem is the 1% when it doesn't. This is the problem with race conditions, 500ms might be ok now, until you change some other timing condition in your system.

I really would advise using a MediaList instead of this.

**Comment by caprica** [comment]
You will have much less problems if you re-use players rather than creating one each time - it also seems you never release your media player anywhere. These things not being done quite right can cloud the issue.

**Comment by caprica** [comment]
I also think you might yet have a pinning references problem, but I don't really have the time or resources to study your code in detail. Your Tester instance goes out of scope on the next loop iteration for example.

**Comment by caprica** [comment]
On your comment about using more than one media player... that is in theory better since invoking methods on one media player should be independent of the other. However, this also is not 100% reliable and crashes can still occur (especially as you increase the number of media players). This particular aspect of your problem is what vlcj-pro was created to solve - i.e. multiple separate media players each in their own JVM.

**Comment by der-lange-christian** [comment]
I wanted to use the vlcj project to produce music in a radio, so is this at all a bad idea with all the not 100% reliable-things ?
- I need there at least 3 players
   + player for current song
   + player for jingles / adverts (while normal song is still running)
   + player to listen in the songs before they go on air

Using the MediaList is a problem with changes in the list isn't it? How long can the list be? And which element is played currently?

**Comment by caprica** [comment]
vlcj is perfectly reliable if you use it the "right" way. I can build an app like you describe reliably enough so that there's no issue. I just explained here possible reasons why you're seeing the problem you described in your issue report and your micro-benchmark.

The essential evaluation of your reported issue is that not being able to callback into VLC is a fact, as is the asynchronous nature of many of LibVLC's API calls, so you have to build your application with those things in mind.

Anyway, you could have flaws in your own code... you create a new media player each iteration, you add an event listener, your object goes out of scope, you don't clean up after yourself, you're likely leaking native resources. This alone could be why you're having problems, and nothing to do with what's being discussed here.

I don't know why you think there are problems with MediaList. You can track the current element by handling events properly. A big advantage of the media list is that each media plays back instantly with barely a break. When playing your own playlist with events there can be a noticeable delay between the end of one media and the start of the next.

For issues like this, there's no simple "just do this" answers. It depends a lot on how you've written your application.

**Comment by caprica** [comment]
I'm going to close this, but reopen #351.
